src/k8s/pkg/k8sd/setup/containerd.go

```diff
@@ -166,23 +167,79 @@ func Containerd(snap snap.Snap, extraContainerdConfig map[string]any, extraArgs
 	return nil
 }
 
-func saveSnapContainerdPaths(s snap.Snap) error {
-	// Write the containerd-related paths to files to properly clean-up on removal.
+// Returns a mapping between the absolute paths of the lockfiles within the
+// k8s snap and the absolute paths of the containerd directory they lock.
+func containerdLockPathsForSnap(s snap.Snap) map[string]string {
 	m := map[string]string{
 		"containerd-socket-path": s.ContainerdSocketDir(),
 		"containerd-config-dir":  s.ContainerdConfigDir(),
 		"containerd-root-dir":    s.ContainerdRootDir(),
 		"containerd-cni-bin-dir": s.CNIBinDir(),
 	}
 
-	for filename, content := range m {
-		if err := utils.WriteFile(filepath.Join(s.LockFilesDir(), filename), []byte(content), 0o600); err != nil {
-			return fmt.Errorf("failed to write %s: %w", filename, err)
+	prefixed := map[string]string{}
+	for k, v := range m {
+		prefixed[filepath.Join(s.LockFilesDir(), k)] = v
+	}
+
+	return prefixed
+}
+
+// Creates the lock files for the containerd directory paths to be used for later cleanup.
+func saveSnapContainerdPaths(s snap.Snap) error {
+	for lockpath, dirpath := range containerdLockPathsForSnap(s) {
+		if err := utils.WriteFile(lockpath, []byte(dirpath), 0o600); err != nil {
+			return fmt.Errorf("failed to write %s: %w", lockpath, err)
 		}
 	}
 	return nil
 }
 
+// Attempts to clean up all containerd directories which were created
+// by the k8s-snap based on the existence of their respective lockfiles
+// located in the directory returned by `s.LockFilesDir()`.
+func TryCleanupContainerdPaths(s snap.Snap) error {
+	log := log.L()
+
+	for lockpath, dirpath := range containerdLockPathsForSnap(s) {
+		// Ensure lockfile exists:
+		if _, err := os.Stat(lockpath); os.IsNotExist(err) {
+			log.Info("WARN: failed to find containerd lockfile, no cleanup will be perfomed", "lockfile", lockpath, "directory", dirpath)
+			continue
+		}
+
+		// Ensure lockfile's contents is the one we expect:
+		lockfile_contents := ""
+		if contents, err := os.ReadFile(lockpath); err != nil {
+			log.Info("WARN: failed to read contents of lockfile", "lockfile", lockpath, "error", err)
+			continue
+		} else {
+			lockfile_contents = string(contents)
+		}
+
+		if lockfile_contents != dirpath {
+			log.Info("WARN: lockfile points to different path than expected", "lockfile", lockpath, "expected", dirpath, "actual", lockfile_contents)
+			continue
+		}
+
+		// Check directory exists before attempting to remove:
+		if _, err := os.Stat(dirpath); os.IsNotExist(err) {
+			log.Info("Containerd directory doesn't exist; skipping cleanup", "directory", dirpath)
+		} else {
+			if err := os.RemoveAll(dirpath); err != nil {
+				log.Info("WARN: failed to remove containerd data directory", "directory", dirpath, "error", err)
+				continue // Avoid removing the lockfile path.
```

Comment:
Isn't cleanup done on a best effort basis? Do we check the lockfile later on at bootstrap time?

Reply:
It is, but IMO it would have been "incorrect" for this code to remove the lockfile path if it had failed to remove the actual path it was pointing too. (IMO either both should be deleted; or none)

> Do we check the lockfile later on at bootstrap time?

AFAIK the `LockFilesDir()` is used only by `k8sd` in this file when setting up the containerd dirs, and [to mark a a node as worker](https://github.com/canonical/k8s-snap/blob/a00846526c10cca25cfe54d2ad1685e558e7251a/src/k8s/pkg/snap/util/node.go#L13) (contents doesn't matter for marking workers, only the file existing does).

These contents of the [lockfiles are only read by the `remove` hook of the k8s-snap](https://github.com/canonical/k8s-snap/blob/a00846526c10cca25cfe54d2ad1685e558e7251a/k8s/lib.sh#L104) (it also neglects to delete the lockfiles after it's done though, unsure whether that's intentional or not...).

Reply:
I agree that removing the lockfile on failure, my thought/question is how should we communicate this to the user since these are just logs. Should we have a check in the bootstrap process so we can point out potentially left over paths? @bschimke95